server/text_generation_server/models/vlm_causal_lm.py

```diff
@@ -409,15 +698,237 @@ def __init__(
+                    seqlen=seqlen,
+                    max_s=max_s,
+                    prefill_cache_indices=None,
+                    lm_head_indices=None,
+                )
+                self.cuda_graphs[bs]["logits"] = logits
+                self.cuda_graphs[bs]["speculative_logits"] = speculative_logits
+        torch.cuda.synchronize()
+
+    def get_vision_embeds(
+        self,
+        pixel_values: torch.Tensor,
+        pixel_attention_mask: torch.Tensor,
+        image_sizes: torch.Tensor,
+        image_grid_thw: torch.Tensor,
+    ):
+        embeds = self.model.get_vision_embeds(
+            pixel_values=pixel_values,
+            pixel_attention_mask=pixel_attention_mask,
+            image_sizes=image_sizes,
+            image_grid_thw=image_grid_thw,
+        )
+        return embeds
+
+    def get_inputs_embeds(
+        self,
+        input_ids: torch.Tensor,
+        vision_embeds: Optional[torch.Tensor] = None,
+    ):
+        return self.model.get_inputs_embeds(
+            input_ids=input_ids,
+            vision_embeds=vision_embeds,
+        )
+
+    def encode_images(self, batch):
+        if batch.pixel_values is not None:
+            device = batch.input_ids.device
+            for request_id, image_id, image_input in batch.pixel_values:
+                pixel_values = image_input["pixel_values"].to(device)
+
+                if "pixel_attention_mask" in image_input:
+                    pixel_attention_mask = image_input["pixel_attention_mask"].to(
+                        device
+                    )
+                else:
+                    pixel_attention_mask = None
+
+                if "image_sizes" in image_input:
+                    image_sizes = image_input["image_sizes"].to(device)
+                else:
+                    image_sizes = None
+
+                if "image_grid_thw" in image_input:
+                    image_grid_thw = image_input["image_grid_thw"].to(device)
+                else:
+                    image_grid_thw = None
+
+                encoder_outputs = self.get_vision_embeds(
+                    pixel_values=pixel_values,
+                    pixel_attention_mask=pixel_attention_mask,
+                    image_sizes=image_sizes,
+                    image_grid_thw=image_grid_thw,
+                )
+                batch.update_encoder_cache(
+                    encoder_outputs,
+                    request_id,
+                    batch.image_positions[request_id][image_id],
+                )
+
+        batch.pixel_values = None
+
+    def get_input_embeddings(self, batch):
+        if batch.has_image_inputs:
+            self.encode_images(batch)
+            vision_embeds = batch.gather_vision_embeds()
+            batch.has_image_inputs = False
+        else:
+            vision_embeds = None
+
+        inputs_embeds = self.get_inputs_embeds(
+            batch.input_ids, vision_embeds=vision_embeds
+        )
+
+        batch.inputs_embeds = inputs_embeds
```

Comment:
The function name is not correct, `get_xx` should always return something.
If you're not returning something but instead acting on some object, the name can be `set_xxx`.

Also let's keep consistency and use `input_embeds` everywhere instead of `input_embeddings`.

Reply:
added consistency in names of method and args